File: src/Tint.Core/Reload/Reloaders.cs
using System.Diagnostics;
using Tint.Core.Themes;
using Tint.Core.Util;

namespace Tint.Core.Reload;

public sealed record ReloadContext(Scheme Scheme, string Wallpaper, string CacheDirectory);

/// <summary>What happened to one app. <see cref="Skipped"/> means "not installed / not running" — not an error.</summary>
public sealed record ReloadResult(string Name, bool Ok, bool Skipped, string Detail);

/// <summary>Tells one app to pick up the new colours.</summary>
public interface IReloader
{
    string Name { get; }

    ReloadResult Reload(ReloadContext context);
}

public static class Reloaders
{
    /// <summary>The reload steps for this OS, in order, ending with the user's hook.</summary>
    public static IReadOnlyList<IReloader> ForCurrentPlatform()
    {
        var list = new List<IReloader>();
        if (OperatingSystem.IsLinux())
        {
            list.Add(new CommandReloader("Hyprland", "hyprctl", ["reload"],
                requires: () => Environment.GetEnvironmentVariable("HYPRLAND_INSTANCE_SIGNATURE") is { Length: > 0 },
                whyNot: "not running inside Hyprland"));
            list.Add(new CommandReloader("kitty", "pkill", ["-USR1", "-x", "kitty"], notRunningExitCode: 1));
            list.Add(new GtkColorSchemeReloader());
            list.Add(new CommandReloader("Firefox (pywalfox)", "pywalfox", ["update"]));
        }

        list.Add(new HookReloader());
        return list;
    }
}

/// <summary>Runs one command, if it's installed.</summary>
public sealed class CommandReloader(
    string name,
    string command,
    string[] arguments,
    Func<bool>? requires = null,
    string? whyNot = null,
    int? notRunningExitCode = null) : IReloader
{
    public string Name => name;

    public ReloadResult Reload(ReloadContext context)
    {
        if (!TintPaths.OnPath(command))
        {
            return new ReloadResult(name, Ok: true, Skipped: true, $"{command} not installed");
        }

        if (requires is not null && !requires())
        {
            return new ReloadResult(name, Ok: true, Skipped: true, whyNot ?? "not applicable");
        }

        try
        {
            ProcessResult result = ProcessRunner.Run(command, arguments);
            if (notRunningExitCode is { } code && result.ExitCode == code)
            {
                return new ReloadResult(name, Ok: true, Skipped: true, "not running");
            }

            return result.Succeeded
                ? new ReloadResult(name, Ok: true, Skipped: false, "reloaded")
                : new ReloadResult(name, Ok: false, Skipped: false, $"{command} exited {result.ExitCode}: {result.StdErr}");
        }
        catch (Exception ex) when (ex is TimeoutException or InvalidOperationException or System.ComponentModel.Win32Exception)
        {
            return new ReloadResult(name, Ok: false, Skipped: false, ex.Message);
        }
    }
}

/// <summary>Tells GTK apps to prefer dark or light, matching the scheme.</summary>
public sealed class GtkColorSchemeReloader : IReloader
{
    public string Name => "GTK dark/light";

    public ReloadResult Reload(ReloadContext context)
    {
        string preference = context.Scheme.Mode == ThemeMode.Dark ? "prefer-dark" : "prefer-light";
        ReloadResult result = new CommandReloader(Name, "gsettings", ["set", "org.gnome.desktop.interface", "color-scheme", preference])
            .Reload(context);

        // gsettings can be installed without GNOME's schemas (common on
        // Hyprland setups): nothing to set, which isn't a failure.
        if (!result.Ok && result.Detail.Contains("No such schema", StringComparison.Ordinal))
        {
            return new ReloadResult(Name, Ok: true, Skipped: true, "GNOME settings schema not installed");
        }

        return result is { Ok: true, Skipped: false } ? result with { Detail = preference } : result;
    }
}

/// <summary>
/// Runs <c>~/.config/tint/hooks/post-apply</c> if it exists and is executable —
/// the place for setup-specific steps (restarting a shell, copying a scheme
/// file somewhere). It gets TINT_WALLPAPER, TINT_MODE and TINT_CACHE.
/// </summary>
public sealed class HookReloader(string? hooksDirectory = null) : IReloader
{
    private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(30);

    public string Name => "post-apply hook";

    public ReloadResult Reload(ReloadContext context)
    {
        string hook = Path.Combine(hooksDirectory ?? TintPaths.Hooks, "post-apply");
        if (!File.Exists(hook))
        {
            return new ReloadResult(Name, Ok: true, Skipped: true, $"no {hook}");
        }

        if (!OperatingSystem.IsWindows() && (File.GetUnixFileMode(hook) & UnixFileMode.UserExecute) == 0)
        {
            return new ReloadResult(Name, Ok: false, Skipped: false, $"{hook} is not executable (chmod +x it)");
        }

        // Output is NOT captured: hooks often start background processes (a
        // shell, a bar), which would keep a captured pipe open and hang tint.
        var info = new ProcessStartInfo(hook) { UseShellExecute = false };
        info.Environment["TINT_WALLPAPER"] = context.Wallpaper;
        info.Environment["TINT_MODE"] = context.Scheme.Mode == ThemeMode.Dark ? "dark" : "light";
        info.Environment["TINT_CACHE"] = context.CacheDirectory;

        try
        {
            using Process process = Process.Start(info) ?? throw new InvalidOperationException("could not start the hook");
            if (!process.WaitForExit(Timeout))
            {
                return new ReloadResult(Name, Ok: false, Skipped: false, $"still running after {Timeout.TotalSeconds:0} s; left it running");
            }

            return process.ExitCode == 0
                ? new ReloadResult(Name, Ok: true, Skipped: false, "ran")
                : new ReloadResult(Name, Ok: false, Skipped: false, $"exited {process.ExitCode}");
        }
        catch (Exception ex) when (ex is InvalidOperationException or System.ComponentModel.Win32Exception)
        {
            return new ReloadResult(Name, Ok: false, Skipped: false, ex.Message);
        }
    }
}

File: Sources/TintCore/Reload/Reloaders.swift
import Foundation

public struct ReloadContext: Sendable {
    public var scheme: Scheme
    public var wallpaper: String
    public var cacheDirectory: String
}

/// What happened to one app. `skipped` means "not installed / not running" — not an error.
public struct ReloadResult: Sendable {
    public var name: String
    public var ok: Bool
    public var skipped: Bool
    public var detail: String

    static func done(_ name: String, _ detail: String) -> ReloadResult { .init(name: name, ok: true, skipped: false, detail: detail) }
    static func skip(_ name: String, _ detail: String) -> ReloadResult { .init(name: name, ok: true, skipped: true, detail: detail) }
    static func fail(_ name: String, _ detail: String) -> ReloadResult { .init(name: name, ok: false, skipped: false, detail: detail) }
}

/// Tells one app to pick up the new colours.
public protocol Reloader: Sendable {
    var name: String { get }
    func reload(_ context: ReloadContext) -> ReloadResult
}

public enum Reloaders {
    /// The reload steps, in order. The user's hook goes last, so it can build
    /// on (or undo) anything the built-in steps did.
    public static func all() -> [any Reloader] {
        [SketchyBarReloader(), BordersReloader(), GhosttyReloader(), HookReloader()]
    }

    /// Runs a command and turns its outcome into a result.
    static func command(_ name: String, _ program: String, _ arguments: [String]) -> ReloadResult {
        do {
            let result = try ProcessRunner.run(program, arguments)
            return result.succeeded ? .done(name, "reloaded") : .fail(name, "\(program) exited \(result.exitCode): \(result.stderr)")
        } catch {
            return .fail(name, "\(error)")
        }
    }

    /// Runs a script without capturing output (it may start background processes).
    static func script(_ name: String, _ path: String, timeout: TimeInterval, environment: [String: String] = [:], ran: String) -> ReloadResult {
        do {
            switch try ProcessRunner.runDetached(path, timeout: timeout, environment: environment) {
            case 0: return .done(name, ran)
            case nil: return .fail(name, "\(path) still running after \(Int(timeout)) s; left it running")
            case let code?: return .fail(name, "\(path) exited \(code)")
            }
        } catch {
            return .fail(name, "couldn't run \(path): \(error.localizedDescription) (is it executable?)")
        }
    }
}

/// SketchyBar re-runs its `sketchybarrc` on `--reload`, so a bar config that
/// reads `~/.cache/wal` picks up the new colours.
public struct SketchyBarReloader: Reloader {
    public let name = "SketchyBar"

    public init() {}

    public func reload(_ context: ReloadContext) -> ReloadResult {
        guard ProcessRunner.isRunning("sketchybar") else { return .skip(name, "not running") }
        return Reloaders.command(name, "sketchybar", ["--reload"])
    }
}

/// JankyBorders. Calling `borders` with options while it runs updates the
/// running instance, so there's no restart. If there's a
/// `~/.config/borders/bordersrc` it is run — it usually reads `colors.sh` and
/// passes the colours on, keeping your own style; otherwise the active border
/// gets blue (color4) and the inactive one color8.
public struct BordersReloader: Reloader {
    public let name = "JankyBorders"
    let bordersrc: String

    public static var defaultBordersrc: String { TintPaths.home + "/.config/borders/bordersrc" }

    public init(bordersrc: String = BordersReloader.defaultBordersrc) { self.bordersrc = bordersrc }

    public func reload(_ context: ReloadContext) -> ReloadResult {
        guard ProcessRunner.isRunning("borders") else { return .skip(name, "not running") }
        if TintPaths.exists(bordersrc) {
            return Reloaders.script(name, bordersrc, timeout: 10, ran: "reloaded (ran bordersrc)")
        }
        return Reloaders.command(name, "borders", [
            "active_color=0xff\(context.scheme[4].strip)",
            "inactive_color=0xff\(context.scheme[8].strip)",
        ])
    }
}

/// Ghostty 1.2+ reloads its config on SIGUSR2. That only changes colours if
/// the config includes tint's theme file, so it checks for that first:
/// `config-file = ~/.cache/wal/colors-ghostty`.
public struct GhosttyReloader: Reloader {
    public static let themeFile = "colors-ghostty"
    public let name = "Ghostty"
    let configFiles: [String]

    public static var defaultConfigFiles: [String] {
        let env = ProcessInfo.processInfo.environment["XDG_CONFIG_HOME"]
        let xdg = (env?.isEmpty == false ? env! : TintPaths.home + "/.config") + "/ghostty"
        let support = TintPaths.home + "/Library/Application Support/com.mitchellh.ghostty"
        return [xdg + "/config", xdg + "/config.ghostty", support + "/config", support + "/config.ghostty"]
    }

    public init(configFiles: [String] = GhosttyReloader.defaultConfigFiles) { self.configFiles = configFiles }

    /// True if some Ghostty config mentions tint's theme file.
    public func usesTintTheme() -> Bool {
        configFiles.contains { (try? String(contentsOfFile: $0, encoding: .utf8))?.contains(GhosttyReloader.themeFile) == true }
    }

    public func reload(_ context: ReloadContext) -> ReloadResult {
        guard ProcessRunner.isRunning("ghostty") else { return .skip(name, "not running") }
        guard usesTintTheme() else {
            return .skip(name, "add `config-file = ~/.cache/wal/\(GhosttyReloader.themeFile)` to your Ghostty config to theme it")
        }
        return Reloaders.command(name, "pkill", ["-USR2", "-x", "ghostty"])
    }
}

/// Runs `~/.config/tint/hooks/post-apply` if it exists and is executable —
/// the place for setup-specific steps (editor themes…). It gets
/// TINT_WALLPAPER, TINT_MODE and TINT_CACHE.
public struct HookReloader: Reloader {
    public let name = "post-apply hook"
    let hooksDirectory: String

    public init(hooksDirectory: String = TintPaths.hooks) { self.hooksDirectory = hooksDirectory }

    public func reload(_ context: ReloadContext) -> ReloadResult {
        let hook = hooksDirectory + "/post-apply"
        guard TintPaths.exists(hook) else { return .skip(name, "no \(hook)") }
        guard FileManager.default.isExecutableFile(atPath: hook) else {
            return .fail(name, "\(hook) is not executable (chmod +x it)")
        }
        return Reloaders.script(name, hook, timeout: 30, environment: [
            "TINT_WALLPAPER": context.wallpaper,
            "TINT_MODE": context.scheme.mode.rawValue,
            "TINT_CACHE": context.cacheDirectory,
        ], ran: "ran")
    }
}
